networkx/algorithms/euler.py

```diff
@@ -434,13 +438,19 @@ def eulerize(G):
         for m, n in combinations(odd_degree_nodes, 2)
     ]
 
-    # use inverse path lengths as edge-weights in a new graph
+    # get the maximum length (in terms of the number of vertices)
+    # of a shortest path between a pair of vertices of odd degree in G
+    max_odd_path_length = max(
+        map(lambda path: len(next(iter(path[1].values()))), odd_deg_pairs_paths)
```

Comment:
Do you see a more appropriate way for computing the length of a path than `len(next(iter(path[1].values())))` ?

Reply:
As long as the "max value" is bigger than any path length, the new weighting scheme should work... right?
All the edge lengths would be longer by the same amount, but the shortest path would still be the shortest.

So we could use 
```python
maxPL = len(G)+1    # this should be much faster too  
... 
                  Gp.add_edge(m, n, weight=maxPL - len(P), path=P)
```

Reply:
> As long as the "max value" is bigger than any path length, the new weighting scheme should work... right?

Yes, we just need the weights to be positive for the max. matching algorithm to output a _perfect_ matching. 
I was just a bit afraid that having unnecessarily larger weights would make [max_weight_matching](https://github.com/networkx/networkx/blob/73afc6768021f46279d05e6b8a8e722bffac1d79/networkx/algorithms/euler.py#L446) invocation slower, but now I see that this doesn't make a difference unless the weights do not fit into 4 bytes which is at least about $10^9$ and therefore shouldn't be an issue.

Then `maxPL = len(G)+1` mainly simplifies the code, since a clear bottleneck here is the computation of all the shortest paths  between odd-degree vertices [here](https://github.com/networkx/networkx/blob/73afc6768021f46279d05e6b8a8e722bffac1d79/networkx/algorithms/euler.py#L433). However, I think that it should be quite easy to introduce [edge-weights](https://github.com/rfulekjames/networkx/compare/fix--for-issue-5952...rfulekjames:networkx:follow-up--for-issue-5952) as a parameter to `eulerize` function in which case we should be computing max. path length.

After looking into this closer, I find the computation of all the shortest paths as it is now sub-optimal in the case when substantially many vertices in the graph have odd degree (as is the case for almost all graphs) in which case floyd_warshall algorithm should be used, right? At any rate, it should be always good to compute shortest paths from one source to multiple targets with a single call of Dijkstra algorithm.
For example, on a 30 x 30 hexagonal lattice `eulerize` runs for more than half an hour on my laptop mainly due to the computation of all the shortest paths.

Reply:
I agree that it would be better/faster most of the time to compute all the targets for any starting node rather than computing the shortest paths between each possible pair of odd degree nodes. There might be cases where this is slower than the current setup (e.g. when the number of even degree nodes is much much greater than the number of odd degree nodes). But a single dijkstra pass should be better in most all cases.  

I notice that we compute all those paths, but then mostly only use their lengths.  In fact the only place we need the paths themselves is to store the path as an attribute on the edge of the new graph.  Can you check the time of a version that computes path_length instead of path?  I suspect the path_length routines are faster than the path routines by enough that we should consider an option to not compute the paths here. But the default should be to have the paths I suspect.  But this whole paragraph is probably not as important as the previous paragraph. If we can reduce the number of graph traversals we need to do that will have the biggest benefit (I think).

Reply:
> Can you check the time of a version that computes path_length instead of path? I suspect the path_length routines are faster than the path routines by enough that we should consider an option to not compute the paths here

On `hexagonal_lattice_graph(20, 20)` and `hexagonal_lattice_graph(30, 30)` the running time improvement if not computing all the shortest paths is about 10% on my laptop (10 vs 9 sec. and 77 vs 70 sec, resp.), but the main benefit of not computing them should be in saving memory.


> There might be cases where this is slower than the current setup (e.g. when the number of even degree nodes is much much greater than the number of odd degree nodes). But a single dijkstra pass should be better in most all cases.

On `hexagonal_lattice_graph(30, 30)`  the running time went down considerably, to roughly 1/30 (if using a single  dijkstra pass from each vertex) of what it is now on my laptop. If odd degree nodes are _localized_ to a small part of graph the way it is implemented now is can be, of course, much better. 

To solve both cases (_localized_ and _non-localized_) at once a function computing shortest path lengths from a single source to multiple targets (without need to always search the whole graph) would be desirable, but I coudn't find such a function in the library. Have I overlooked anything? If there is no such function, perhaps, adding it would be beneficial also for different use cases.

Alternatively one can introduce a parameter in the `eulerize` function that would switch between the strategies as for how to compute shortest path lengths.